File: Cpp/0024_swap_nodes_in_pairs.cpp
// https://leetcode.cn/problems/swap-nodes-in-pairs/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode *dummy = new ListNode(), *cur = new ListNode();
        dummy->next = cur;
        cur->next = head;
        while (head != nullptr && head->next != nullptr) {
            ListNode *tmp = head->next;
            cur->next = tmp;
            head->next = tmp->next;
            tmp->next = head;
            cur = head;
            head = head->next;
        }
        return dummy->next->next;
    }
};

File: Python3/0024_swap_nodes_in_pairs.py
# https://leetcode.cn/problems/swap-nodes-in-pairs/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        cur = ListNode()
        cur.next = head
        dummy.next = cur
        while head and head.next:
            tmp = head.next
            cur.next = tmp
            head.next = tmp.next
            tmp.next = head
            cur = head
            head = head.next
        return dummy.next.next